modules/post/windows/gather/credentials/packrat_credentials.rb

```diff
@@ -0,0 +1,2230 @@
+require 'rex'
+require 'msf/core/post/windows/user_profiles'
+require 'json'
+require 'nokogiri'
+require 'sqlite3'
+
+class MetasploitModule < Msf::Post
+
+  include Msf::Post::File
+  include Msf::Post::Windows::UserProfiles
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name' => 'Windows Gather Application Artifacts (PackRat)',
+      'Description' => %q{PackRat gathers artifacts of various categories from a large number of applications.
+       Artifacts include: 12 browsers, 13 chat/IM/IRC applications, 6 email clients, and 1 game.
+       Credentials are then extracted from the artifacts. The use case for this post-exploitation module is
+       to specify the types of
+       artifacts you are interested in, to gather the relevant files depending on your aims.
+       Please refer to the options for a full list of filter categories.},
+      'License' => MSF_LICENSE,
+      'Author' =>
+        [
+          'Daniel Hallsworth', # Leeds Beckett University student
+          'Barwar Salim M', # Leeds Beckett University student
+          'Z. Cliffe Schreuders (http://z.cliffe.schreuders.org)' # Leeds Beckett University lecturer
+        ],
+      'Platform' => %w{win},
+      'SessionTypes' => ['meterpreter']
+      ))
+
+    register_options(
+      [
+        OptRegexp.new('REGEX', [false, 'Match a regular expression', '^password']),
+        OptBool.new('STORE_LOOT', [false, 'Store artifacts into loot database (otherwise, only download)', 'true']),
+        # enumerates the options based on the artifacts that are defined below
+        OptEnum.new('APPCATEGORY', [false, 'Category of applications to gather from', 'All', APPLICATION_ARRAY.map {|x| x[:category]}.uniq.unshift('All')]),
+        OptEnum.new('APPLICATION', [false, 'Specify application to gather from', 'All', APPLICATION_ARRAY.map {|x| x[:application]}.uniq.unshift('All')]),
+        OptEnum.new('ARTEFACTS', [false, 'Type of artifacts to collect', 'All', APPLICATION_ARRAY.map {|x| x[:filetypes]}.uniq.unshift('All')]),
+      ])
+  end
+
+  # this associative array defines the artifacts known to PackRat
+  APPLICATION_ARRAY =
```

Comment:
Longest hash ever. Can we put this in a JSON file?

Reply:
I worked a bit at this yesterday, and I think it's doable so long as we process the regexes correctly. They'd be converted to strings and would need to be converted back.

Reply:
would it work if I saved the regular expressions as a string in the application array e.g. 
:regex => ["/password.*/i", "username.*/i"]

Reply:
When you do `to_json` on the hash (associative array), regexes will get converted into regex strings. You can then use `Regexp.new` or `Regexp.compile` on the strings to return them to their regex form. I haven't tested that much, but I think that's how it'd work.

Reply:
```
[1] pry(#<Msf::Modules::Post__Windows__Gather__Credentials__Packrat_credentials::MetasploitModule>)> APPLICATION_ARRAY == JSON.parse(APPLICATION_ARRAY.to_json)
=> true
[2] pry(#<Msf::Modules::Post__Windows__Gather__Credentials__Packrat_credentials::MetasploitModule>)>
```

Reply:
I prefer `JSON.pretty_generate`. How about I send you a PR? Here's what it looks like so far:

```
[2] pry(#<Msf::Modules::Post__Windows__Gather__Credentials__Packrat_credentials::MetasploitModule>)> Dir.mkdir('data/packrat') && File.write('data/packrat/artifacts.json', JSON.pretty_generate(APPLICATION_ARRAY))
=> 59907
[3] pry(#<Msf::Modules::Post__Windows__Gather__Credentials__Packrat_credentials::MetasploitModule>)>
```

```
wvu@kharak:/rapid7/metasploit-framework/data/packrat:HEAD$ file artifacts.json
artifacts.json: ASCII text
wvu@kharak:/rapid7/metasploit-framework/data/packrat:HEAD$ head -20 !$
head -20 artifacts.json
[
  {
    "application": "incredimail",
    "category": "emails",
    "file_artifact": [
      {
        "filetypes": "email_logs",
        "path": "LocalAppData",
        "dir": "IM",
        "artifact": "msg.iml",
        "description": "IncrediMail sent and received emails",
        "credential_type": "text",
        "regex_search": [
          {
            "extraction_description": "Searches for credentials (USERNAMES/PASSWORDS)",
            "extraction_type": "credentials",
            "regex": [
              "(?i-mx:password.*)",
              "(?i-mx:username.*)"
            ]
wvu@kharak:/rapid7/metasploit-framework/data/packrat:HEAD$
```

Reply:
I'm preparing a PR for you. There are a number of other changes within.

Reply:
https://github.com/danhallsworth1/metasploit-framework/pull/1

Reply:
Thankyou i'll get these issues fixed

Reply:
@wvu-r7 Hi i've made the  changes now and merged the branches can you see it?

Reply:
Yes, I can! Thank you.